src/app/clusters/time-synchronization-server/time-synchronization-delegate.h

```diff
@@ -109,7 +109,9 @@ class Delegate
     /**
      * @brief Signals application that UTCTime has changed through the timesync cluster.
      */
-    virtual void UTCTimeAvailabilityChanged(uint64_t time) = 0;
+    virtual void UTCTimeAvailabilityChanged(uint64_t time)                                         = 0;
+    virtual void TrustedTimeSourceAvailabilityChanged(bool available, GranularityEnum granularity) = 0;
+    virtual void NotifyTimeFailure()                                                               = 0;
```

Comment:
how about passing the underlying CHIP_ERROR to this method, then the default implementation could log or do something, otherwise we do not know much besides "some failure occurred" which I would generally not know what to do with ... what do we expect implementations to do when they receive this? If I get a failure at some point, how do I know when the failure is gone? Some documentation on this would be useful.

Reply:
This is meant to be used as "time cannot be fetched from TrustedTimeSource". When time is available again, UTCTimeAvailabilityChanged() would be called.
But, passing why it failed could be done - just didn't see the importance of it for my use case.

Reply:
I think we should pass it in.

please also add documentation describing lifetime: that we have an error and you may get a UTCTimeAvailabilityChanged when the error condition goes away (and also describe if this is a backgroup automatic operation or if something needs to handle NotifyTimeFailurte in some way to trigger a retry)

Reply:
Had a look at this again and I only use this when we emit the time failure event. Therefore, I cannot put any error as a parameter here. I can add a description of what it means. Is there a way to be notified of events in the SDK internally for apps? If that exists I can remove this function.

Reply:
You have to plumb this more, however you do have different chip-errors. I see at least in the code:

```cpp
    if (GetDefaultNtp(span) != CHIP_NO_ERROR)
    {
        emitTimeFailureEvent(kRootEndpointId);
        return;
    }
    if (span.size() > kMaxDefaultNTPSize)
    {
        emitTimeFailureEvent(kRootEndpointId);
        return;
    }
    if (GetDelegate()->UpdateTimeUsingNTPFallback(span, &mOnFallbackNTPCompletion) != CHIP_NO_ERROR)
    {
        emitTimeFailureEvent(kRootEndpointId);
    }
```

so you have:
   - CHIP_ERROR from GetDefaultNtp
   - some custom error that should represent overly large span
   - CHIP_ERROR from UpdateTimeUsingNTPFallback

Right now this error is actually LOST so one cannot differentiate the type of error it gets and it is not even logged ... so if a time failure occurs I do not know if the span get failed, the span size is invalid or if the update fallback failed. We should not lose errors and when an error callback is called, the callback should be told what error happened to the best of our abilities. messages of "something failed" are hard to debug.

Reply:
I don't think this is a comment relevant to this PR. This code existed before and isn't even mine. Yes we should improve it but not in this PR.